File: app/src/main/java/com/bscan/model/graph/PropertyValue.kt
package com.bscan.model.graph

import com.bscan.model.graph.entities.TrackingMode
import java.time.LocalDateTime
import java.time.LocalDate

/**
 * Type-safe property value container supporting multiple data types.
 * Handles serialization and type conversion for graph entity/edge properties.
 */
sealed class PropertyValue {
    
    abstract val type: PropertyType
    abstract val rawValue: Any?
    
    data class StringValue(override val rawValue: String) : PropertyValue() {
        override val type = PropertyType.STRING
    }
    
    data class IntValue(override val rawValue: Int) : PropertyValue() {
        override val type = PropertyType.INT
    }
    
    data class LongValue(override val rawValue: Long) : PropertyValue() {
        override val type = PropertyType.LONG
    }
    
    data class FloatValue(override val rawValue: Float) : PropertyValue() {
        override val type = PropertyType.FLOAT
    }
    
    data class DoubleValue(override val rawValue: Double) : PropertyValue() {
        override val type = PropertyType.DOUBLE
    }
    
    data class BooleanValue(override val rawValue: Boolean) : PropertyValue() {
        override val type = PropertyType.BOOLEAN
    }
    
    data class DateTimeValue(override val rawValue: LocalDateTime) : PropertyValue() {
        override val type = PropertyType.DATETIME
    }
    
    data class DateValue(override val rawValue: LocalDate) : PropertyValue() {
        override val type = PropertyType.DATE
    }
    
    data class ListValue(override val rawValue: List<*>) : PropertyValue() {
        override val type = PropertyType.LIST
    }
    
    data class MapValue(override val rawValue: Map<String, *>) : PropertyValue() {
        override val type = PropertyType.MAP
    }
    
    data class BytesValue(override val rawValue: ByteArray) : PropertyValue() {
        override val type = PropertyType.BYTES
        
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false
            other as BytesValue
            return rawValue.contentEquals(other.rawValue)
        }
        
        override fun hashCode(): Int = rawValue.contentHashCode()
    }
    
    data class QuantityValue(override val rawValue: Quantity) : PropertyValue() {
        override val type = PropertyType.QUANTITY
    }
    
    object NullValue : PropertyValue() {
        override val type = PropertyType.NULL
        override val rawValue: Nothing? = null
    }
    
    /**
     * Get the value with type safety
     */
    inline fun <reified T> getValue(): T? {
        return when (T::class) {
            String::class -> if (type == PropertyType.STRING) rawValue as T else null
            Int::class -> if (type == PropertyType.INT) rawValue as T else null
            Long::class -> if (type == PropertyType.LONG) rawValue as T else null
            Float::class -> if (type == PropertyType.FLOAT) rawValue as T else null
            Double::class -> if (type == PropertyType.DOUBLE) rawValue as T else null
            Boolean::class -> if (type == PropertyType.BOOLEAN) rawValue as T else null
            LocalDateTime::class -> if (type == PropertyType.DATETIME) rawValue as T else null
            LocalDate::class -> if (type == PropertyType.DATE) rawValue as T else null
            List::class -> if (type == PropertyType.LIST) rawValue as T else null
            Map::class -> if (type == PropertyType.MAP) rawValue as T else null
            ByteArray::class -> if (type == PropertyType.BYTES) rawValue as T else null
            Quantity::class -> if (type == PropertyType.QUANTITY) rawValue as T else null
            else -> null
        }
    }
    
    /**
     * Convert to string representation
     */
    fun asString(): String {
        return when (this) {
            is StringValue -> rawValue
            is IntValue -> rawValue.toString()
            is LongValue -> rawValue.toString()
            is FloatValue -> rawValue.toString()
            is DoubleValue -> rawValue.toString()
            is BooleanValue -> rawValue.toString()
            is DateTimeValue -> rawValue.toString()
            is DateValue -> rawValue.toString()
            is ListValue -> rawValue.toString()
            is MapValue -> rawValue.toString()
            is BytesValue -> rawValue.joinToString("") { "%02x".format(it) }
            is QuantityValue -> rawValue.toString()
            is NullValue -> "null"
        }
    }
    
    companion object {
        /**
         * Create PropertyValue from any supported type
         */
        fun create(value: Any?): PropertyValue {
            return when (value) {
                null -> NullValue
                is String -> StringValue(value)
                is Int -> IntValue(value)
                is Long -> LongValue(value)
                is Float -> FloatValue(value)
                is Double -> DoubleValue(value)
                is Boolean -> BooleanValue(value)
                is LocalDateTime -> DateTimeValue(value)
                is LocalDate -> DateValue(value)
                is List<*> -> ListValue(value)
                is Map<*, *> -> MapValue(value as Map<String, *>)
                is ByteArray -> BytesValue(value)
                is Quantity -> QuantityValue(value)
                else -> StringValue(value.toString())
            }
        }
        
        /**
         * Create PropertyValue from string with type hint
         */
        fun fromString(value: String, type: PropertyType): PropertyValue {
            return when (type) {
                PropertyType.STRING -> StringValue(value)
                PropertyType.INT -> IntValue(value.toIntOrNull() ?: 0)
                PropertyType.LONG -> LongValue(value.toLongOrNull() ?: 0L)
                PropertyType.FLOAT -> FloatValue(value.toFloatOrNull() ?: 0f)
                PropertyType.DOUBLE -> DoubleValue(value.toDoubleOrNull() ?: 0.0)
                PropertyType.BOOLEAN -> BooleanValue(value.toBooleanStrictOrNull() ?: false)
                PropertyType.DATETIME -> DateTimeValue(LocalDateTime.parse(value))
                PropertyType.DATE -> DateValue(LocalDate.parse(value))
                PropertyType.BYTES -> BytesValue(value.chunked(2).map { it.toInt(16).toByte() }.toByteArray())
                PropertyType.QUANTITY -> QuantityValue(Quantity.fromString(value))
                PropertyType.LIST -> StringValue(value)  // JSON parsing would go here
                PropertyType.MAP -> StringValue(value)   // JSON parsing would go here
                PropertyType.NULL -> NullValue
            }
        }
    }
}

/**
 * Supported property types
 */
enum class PropertyType {
    STRING,
    INT,
    LONG,
    FLOAT,
    DOUBLE,
    BOOLEAN,
    DATETIME,
    DATE,
    LIST,
    MAP,
    BYTES,
    QUANTITY,
    NULL
}

/**
 * Property schema definition for validation and UI generation
 */
data class PropertySchema(
    val key: String,
    val type: PropertyType,
    val required: Boolean = false,
    val defaultValue: PropertyValue? = null,
    val validation: PropertyValidation? = null,
    val displayName: String = key,
    val description: String = "",
    val category: String = "general",
    val hidden: Boolean = false,
    val readOnly: Boolean = false
)

/**
 * Property validation rules
 */
data class PropertyValidation(
    val minValue: Number? = null,
    val maxValue: Number? = null,
    val minLength: Int? = null,
    val maxLength: Int? = null,
    val pattern: Regex? = null,
    val allowedValues: Set<String>? = null,
    val customValidator: ((PropertyValue) -> ValidationResult)? = null
)

/**
 * Interface for quantities with units and discrete/continuous tracking modes
 */
interface Quantity {
    val value: Number
    val unit: String
    val trackingMode: TrackingMode
    
    override fun toString(): String
    
    companion object {
        /**
         * Parse a quantity from string representation (e.g., "1000g", "212.5g-discrete")
         */
        fun fromString(str: String): Quantity {
            val parts = str.split("-")
            val quantityPart = parts[0]
            val trackingMode = if (parts.size > 1) {
                when (parts[1].lowercase()) {
                    "discrete" -> TrackingMode.DISCRETE
                    "continuous" -> TrackingMode.CONTINUOUS
                    else -> TrackingMode.CONTINUOUS
                }
            } else TrackingMode.CONTINUOUS
            
            // Extract numeric value and unit
            val numericPart = quantityPart.takeWhile { it.isDigit() || it == '.' }
            val unitPart = quantityPart.drop(numericPart.length)
            
            val value = numericPart.toDoubleOrNull() ?: 0.0
            
            return if (trackingMode == TrackingMode.DISCRETE) {
                DiscreteQuantity(value.toInt(), unitPart.ifEmpty { "units" })
            } else {
                ContinuousQuantity(value, unitPart.ifEmpty { "units" })
            }
        }
    }
}

/**
 * Discrete quantity for countable items
 */
data class DiscreteQuantity(
    override val value: Int,
    override val unit: String
) : Quantity {
    override val trackingMode = TrackingMode.DISCRETE
    
    override fun toString(): String = "${value}${unit}-discrete"
    
    /**
     * Add discrete quantities
     */
    operator fun plus(other: DiscreteQuantity): DiscreteQuantity {
        require(unit == other.unit) { "Cannot add quantities with different units: $unit vs ${other.unit}" }
        return copy(value = value + other.value)
    }
    
    /**
     * Subtract discrete quantities
     */
    operator fun minus(other: DiscreteQuantity): DiscreteQuantity {
        require(unit == other.unit) { "Cannot subtract quantities with different units: $unit vs ${other.unit}" }
        return copy(value = value - other.value)
    }
}

/**
 * Continuous quantity for measurable values
 */
data class ContinuousQuantity(
    override val value: Double,
    override val unit: String
) : Quantity {
    override val trackingMode = TrackingMode.CONTINUOUS
    
    override fun toString(): String = "${value}${unit}"
    
    /**
     * Add continuous quantities
     */
    operator fun plus(other: ContinuousQuantity): ContinuousQuantity {
        require(unit == other.unit) { "Cannot add quantities with different units: $unit vs ${other.unit}" }
        return copy(value = value + other.value)
    }
    
    /**
     * Subtract continuous quantities
     */
    operator fun minus(other: ContinuousQuantity): ContinuousQuantity {
        require(unit == other.unit) { "Cannot subtract quantities with different units: $unit vs ${other.unit}" }
        return copy(value = value - other.value)
    }
    
    /**
     * Multiply by scalar
     */
    operator fun times(factor: Double): ContinuousQuantity {
        return copy(value = value * factor)
    }
    
    /**
     * Divide by scalar
     */
    operator fun div(divisor: Double): ContinuousQuantity {
        require(divisor != 0.0) { "Cannot divide by zero" }
        return copy(value = value / divisor)
    }
}

File: src/models/entities/PropertyValue.ts
/**
 * Type-safe property value container supporting multiple data types.
 * Handles serialization and type conversion for graph entity/edge properties.
 */

import { TrackingMode } from './types';

/**
 * Supported property types
 */
export enum PropertyType {
  STRING = 'STRING',
  INT = 'INT',
  LONG = 'LONG',
  FLOAT = 'FLOAT',
  DOUBLE = 'DOUBLE',
  BOOLEAN = 'BOOLEAN',
  DATETIME = 'DATETIME',
  DATE = 'DATE',
  LIST = 'LIST',
  MAP = 'MAP',
  BYTES = 'BYTES',
  QUANTITY = 'QUANTITY',
  NULL = 'NULL'
}

/**
 * Interface for quantities with units and discrete/continuous tracking modes
 */
export interface Quantity {
  readonly value: number;
  readonly unit: string;
  readonly trackingMode: TrackingMode;
  toString(): string;
}

/**
 * Discrete quantity for countable items
 */
export class DiscreteQuantity implements Quantity {
  readonly trackingMode = TrackingMode.DISCRETE;

  constructor(
    public readonly value: number,
    public readonly unit: string
  ) {
    if (!Number.isInteger(value)) {
      throw new Error('Discrete quantity value must be an integer');
    }
  }

  toString(): string {
    return `${this.value}${this.unit}-discrete`;
  }

  add(other: DiscreteQuantity): DiscreteQuantity {
    if (this.unit !== other.unit) {
      throw new Error(`Cannot add quantities with different units: ${this.unit} vs ${other.unit}`);
    }
    return new DiscreteQuantity(this.value + other.value, this.unit);
  }

  subtract(other: DiscreteQuantity): DiscreteQuantity {
    if (this.unit !== other.unit) {
      throw new Error(`Cannot subtract quantities with different units: ${this.unit} vs ${other.unit}`);
    }
    return new DiscreteQuantity(this.value - other.value, this.unit);
  }
}

/**
 * Continuous quantity for measurable values
 */
export class ContinuousQuantity implements Quantity {
  readonly trackingMode = TrackingMode.CONTINUOUS;

  constructor(
    public readonly value: number,
    public readonly unit: string
  ) {}

  toString(): string {
    return `${this.value}${this.unit}`;
  }

  add(other: ContinuousQuantity): ContinuousQuantity {
    if (this.unit !== other.unit) {
      throw new Error(`Cannot add quantities with different units: ${this.unit} vs ${other.unit}`);
    }
    return new ContinuousQuantity(this.value + other.value, this.unit);
  }

  subtract(other: ContinuousQuantity): ContinuousQuantity {
    if (this.unit !== other.unit) {
      throw new Error(`Cannot subtract quantities with different units: ${this.unit} vs ${other.unit}`);
    }
    return new ContinuousQuantity(this.value - other.value, this.unit);
  }

  multiply(factor: number): ContinuousQuantity {
    return new ContinuousQuantity(this.value * factor, this.unit);
  }

  divide(divisor: number): ContinuousQuantity {
    if (divisor === 0) {
      throw new Error('Cannot divide by zero');
    }
    return new ContinuousQuantity(this.value / divisor, this.unit);
  }
}

/**
 * Quantity factory methods
 */
export const QuantityFactory = {
  /**
   * Parse a quantity from string representation (e.g., "1000g", "212.5g-discrete")
   */
  fromString(str: string): Quantity {
    const parts = str.split('-');
    const quantityPart = parts[0];
    const trackingMode = parts.length > 1 && parts[1].toLowerCase() === 'discrete'
      ? TrackingMode.DISCRETE
      : TrackingMode.CONTINUOUS;

    // Extract numeric value and unit using regex
    const match = quantityPart.match(/^([0-9]*\.?[0-9]+)(.*)$/);
    if (!match) {
      throw new Error(`Invalid quantity format: ${str}`);
    }

    const value = parseFloat(match[1]);
    const unit = match[2] || 'units';

    return trackingMode === TrackingMode.DISCRETE
      ? new DiscreteQuantity(Math.round(value), unit)
      : new ContinuousQuantity(value, unit);
  },

  discrete(value: number, unit: string): DiscreteQuantity {
    return new DiscreteQuantity(value, unit);
  },

  continuous(value: number, unit: string): ContinuousQuantity {
    return new ContinuousQuantity(value, unit);
  }
};

/**
 * Type-safe property value container
 */
export abstract class PropertyValue {
  abstract readonly type: PropertyType;
  abstract readonly rawValue: any;

  /**
   * Get the value with type safety
   */
  abstract getValue<T = any>(): T | undefined;

  /**
   * Convert to string representation
   */
  abstract asString(): string;

  /**
   * Create PropertyValue from any supported type
   */
  static create(value: any): PropertyValue {
    if (value === null || value === undefined) {
      return new NullValue();
    }
    
    if (typeof value === 'string') {
      return new StringValue(value);
    }
    
    if (typeof value === 'number') {
      return Number.isInteger(value) ? new IntValue(value) : new DoubleValue(value);
    }
    
    if (typeof value === 'boolean') {
      return new BooleanValue(value);
    }
    
    if (value instanceof Date) {
      return new DateTimeValue(value);
    }
    
    if (Array.isArray(value)) {
      return new ListValue(value);
    }
    
    if (value instanceof Uint8Array || value instanceof ArrayBuffer) {
      return new BytesValue(new Uint8Array(value));
    }
    
    if (typeof value === 'object' && value !== null && 'value' in value && 'unit' in value && 'trackingMode' in value) {
      return new QuantityValue(value as Quantity);
    }
    
    if (typeof value === 'object' && value !== null) {
      return new MapValue(value as Record<string, any>);
    }
    
    // Fallback to string
    return new StringValue(String(value));
  }

  /**
   * Create PropertyValue from string with type hint
   */
  static fromString(value: string, type: PropertyType): PropertyValue {
    switch (type) {
      case PropertyType.STRING:
        return new StringValue(value);
      case PropertyType.INT:
        return new IntValue(parseInt(value, 10) || 0);
      case PropertyType.DOUBLE:
      case PropertyType.FLOAT:
        return new DoubleValue(parseFloat(value) || 0);
      case PropertyType.BOOLEAN:
        return new BooleanValue(value.toLowerCase() === 'true');
      case PropertyType.DATETIME:
        return new DateTimeValue(new Date(value));
      case PropertyType.DATE:
        return new DateValue(new Date(value));
      case PropertyType.BYTES:
        const bytes = value.match(/.{1,2}/g)?.map(byte => parseInt(byte, 16)) || [];
        return new BytesValue(new Uint8Array(bytes));
      case PropertyType.QUANTITY:
        return new QuantityValue(QuantityFactory.fromString(value));
      case PropertyType.LIST:
        try {
          return new ListValue(JSON.parse(value));
        } catch {
          return new ListValue([value]);
        }
      case PropertyType.MAP:
        try {
          return new MapValue(JSON.parse(value));
        } catch {
          return new MapValue({ value });
        }
      case PropertyType.NULL:
        return new NullValue();
      default:
        return new StringValue(value);
    }
  }
}

export class StringValue extends PropertyValue {
  readonly type = PropertyType.STRING;

  constructor(public readonly rawValue: string) {
    super();
  }

  getValue<T = any>(): T | undefined {
    return this.rawValue as any;
  }

  asString(): string {
    return this.rawValue;
  }
}

export class IntValue extends PropertyValue {
  readonly type = PropertyType.INT;

  constructor(public readonly rawValue: number) {
    super();
    if (!Number.isInteger(rawValue)) {
      throw new Error('IntValue requires an integer');
    }
  }

  getValue<T = any>(): T | undefined {
    return this.rawValue as any;
  }

  asString(): string {
    return this.rawValue.toString();
  }
}

export class DoubleValue extends PropertyValue {
  readonly type = PropertyType.DOUBLE;

  constructor(public readonly rawValue: number) {
    super();
  }

  getValue<T = any>(): T | undefined {
    return this.rawValue as any;
  }

  asString(): string {
    return this.rawValue.toString();
  }
}

export class BooleanValue extends PropertyValue {
  readonly type = PropertyType.BOOLEAN;

  constructor(public readonly rawValue: boolean) {
    super();
  }

  getValue<T = any>(): T | undefined {
    return this.rawValue as any;
  }

  asString(): string {
    return this.rawValue.toString();
  }
}

export class DateTimeValue extends PropertyValue {
  readonly type = PropertyType.DATETIME;

  constructor(public readonly rawValue: Date) {
    super();
  }

  getValue<T = any>(): T | undefined {
    return this.rawValue as any;
  }

  asString(): string {
    return this.rawValue.toISOString();
  }
}

export class DateValue extends PropertyValue {
  readonly type = PropertyType.DATE;

  constructor(public readonly rawValue: Date) {
    super();
  }

  getValue<T = any>(): T | undefined {
    return this.rawValue as any;
  }

  asString(): string {
    return this.rawValue.toISOString().split('T')[0];
  }
}

export class ListValue extends PropertyValue {
  readonly type = PropertyType.LIST;

  constructor(public readonly rawValue: any[]) {
    super();
  }

  getValue<T = any>(): T | undefined {
    return this.rawValue as any;
  }

  asString(): string {
    return JSON.stringify(this.rawValue);
  }
}

export class MapValue extends PropertyValue {
  readonly type = PropertyType.MAP;

  constructor(public readonly rawValue: Record<string, any>) {
    super();
  }

  getValue<T = any>(): T | undefined {
    return this.rawValue as any;
  }

  asString(): string {
    return JSON.stringify(this.rawValue);
  }
}

export class BytesValue extends PropertyValue {
  readonly type = PropertyType.BYTES;

  constructor(public readonly rawValue: Uint8Array) {
    super();
  }

  getValue<T = any>(): T | undefined {
    return this.rawValue as any;
  }

  asString(): string {
    return Array.from(this.rawValue)
      .map(byte => byte.toString(16).padStart(2, '0'))
      .join('');
  }
}

export class QuantityValue extends PropertyValue {
  readonly type = PropertyType.QUANTITY;

  constructor(public readonly rawValue: Quantity) {
    super();
  }

  getValue<T = any>(): T | undefined {
    return this.rawValue as any;
  }

  asString(): string {
    return this.rawValue.toString();
  }
}

export class NullValue extends PropertyValue {
  readonly type = PropertyType.NULL;
  readonly rawValue = null;

  getValue<T = any>(): T | undefined {
    return undefined;
  }

  asString(): string {
    return 'null';
  }
}